msprime/mutations.py

```diff
@@ -56,30 +56,71 @@
 ]
 
 
-# NOTE we're doing this hack of monkey-patching asdict onto the
-# MutationModel subclasses because of a bug in sphinx, where it's
-# not possible to have muliple inheritance with mocked out classes.
-def safe_asdict(self):
-    # This version of asdict makes sure that we have sufficient parameters
-    # to call the contructor and recreate the class. However, this means
-    # that subclasses *must* have an instance variable of the same name.
-    # This is essential for Provenance round-tripping to work.
-    return {
-        key: getattr(self, key)
-        for key in inspect.signature(self.__init__).parameters.keys()
-        if hasattr(self, key)
+def mutation_model_factory(model):
+    """
+    Returns a mutation model corresponding to the specified model.
+    - If model is None, the default mutation model is returned.
+    - If model is a string, return the corresponding model instance.
+    - If model is an instance of MutationModel, return a copy of it.
+    - Otherwise raise a type error.
+    """
+
+    # String names here should match Seq-Gen where possible
+    model_map = {
+        # "slim": SlimMutationModel(), Needs type argument so can't be string init'd
+        "infinite_alleles": InfiniteAllelesMutationModel(),
+        "binary": BinaryMutationModel(),
+        "jc69": JC69MutationModel(),
+        "hky": HKYMutationModel(),
+        "f84": F84MutationModel(),
+        # "gtr": GTRMutationModel(), Needs relative_rates argument, so can't be
+        # string init'd
+        "blosum62": BLOSUM62MutationModel(),
+        "pam": PAMMutationModel(),
     }
 
+    if model is None:
+        model_instance = BinaryMutationModel()
+    elif isinstance(model, str):
+        lower_model = model.lower()
+        if lower_model not in model_map:
+            raise ValueError(
+                "Model '{}' unknown. Choose from {}".format(
+                    model, list(model_map.keys())
+                )
+            )
+        model_instance = model_map[lower_model]
+    elif not isinstance(model, MutationModel):
+        raise TypeError(
+            "Mutation model must be a string or an instance of MutationModel"
+        )
+    else:
+        model_instance = model
+    return model_instance
+
 
-class MatrixMutationModel(_msprime.MatrixMutationModel):
+class MutationModel:
+    def asdict(self):
+        # This version of asdict makes sure that we have sufficient parameters
+        # to call the constructor and recreate the class. However, this means
+        # that subclasses *must* have an instance variable of the same name.
+        # This is essential for Provenance round-tripping to work.
```

Comment:
I don't quite understand this comment - "of the same name" as what?

Reply:
This was here before this PR, but I've tried to clarify in https://github.com/tskit-dev/msprime/pull/1176/commits/ea2547fe7e2da8cf8f8070e8e985118c7daa8e5b